students/johnn/mailroom/tests/test_mailroom.py

```diff
@@ -1,32 +1,125 @@
 #!/usr/bin/env python3
 
-import pytest
+"""
+test script for mailroom
+"""
+
 import io
 import os
-from pprint import pprint
+from unittest.mock import (patch, MagicMock)
+import pytest
 from mailroom import security
 from mailroom.donors import Donor, Donors, load_donor_file, save_donor_file
-from mailroom.ui import print_thank_you, print_lines, thank_all_donors, list_donors
+from mailroom.ui import (print_thank_you,
+                         print_lines,
+                         thank_all_donors,
+                         list_donors,
+                         safe_input,
+                         user_login,
+                         user_logout,
+                         get_donation_amount,
+                         data_entry,
+                         main)
+
+
+# define a default user so we can make objects
+security.user = "Test User"
+
+
+def test_user_logout():
+    """ test the logout routine """
+    assert security.user == "Test User"
+    user_logout()
+    assert security.user is None
+
+
+# define a default user so we can make objects
+security.user = "Test User"
+
+safe_input_normal = MagicMock(return_value="m0cked")
+@patch("builtins.input", safe_input_normal)
+def test_safe_input_normal():
+    """ test the safe_input routine """
+    returned = safe_input()
+    assert returned == "m0cked"
+
+
+safe_input_error = MagicMock(return_value="m0cked")
+@patch("builtins.input", safe_input_error)
+def test_safe_input_error():
+    """ test the safe_input routine """
+    safe_input_error.side_effect = EOFError
+    returned = safe_input()
+    assert returned == ""
+
+
+deck_user_login = MagicMock(return_value="Joe User")
+@patch("builtins.input", deck_user_login)
+def test_user_login():
+    """ test the login routine """
+    user_login()
+    assert security.user == "Joe User"
+
+
+get_donation_normal = MagicMock(return_value="101")
+@patch("builtins.input", get_donation_normal)
+def test_get_donation_normal():
+    """ test the donation input routine """
+    amount = get_donation_amount(Donor(full_name="Test Donor4"))
+    assert amount == 101
+
+
+deck_donation_error = MagicMock(side_effect=["foobar", "999"])
+@patch("builtins.input", deck_donation_error)
+def test_donor_entry_error():
+    """
+    test donation error
+
+    we enter invalid data (string) and then are re-prompted
+    """
+    amount = get_donation_amount(Donor(full_name="Test Donor5"))
+    assert amount == 999
+
+
+get_donations_quit = MagicMock(return_value="q")
+@patch("builtins.input", get_donations_quit)
+def test_get_donation_quit():
+    """ test the donation input routine """
+    amount = get_donation_amount(Donor(full_name="Test Donor5"))
+    assert amount is None
+
```

Comment:
I would be explicit in your tests: If quit successfully from get_donation_amount, None should be returned.